crates/tx5-go-pion/src/data_chan.rs

```diff
@@ -1,106 +1,209 @@
+        if let Some(strong) = self.0.upgrade() {
+            let mut tmp = Err(err.clone());
+
+            {
+                let mut lock = strong.lock().unwrap();
+                if let Ok(core) = &mut *lock {
+                    core.close(err);
+                }
+                std::mem::swap(&mut *lock, &mut tmp);
             }
-            cb
-        });
-        DataChannel(data_chan_id)
+
+            // make sure the above lock is released before this is dropped
+            drop(tmp);
+        }
+    }
+
+    pub fn get_recv_limit(&self) -> Result<Arc<tokio::sync::Semaphore>> {
+        data_chan_weak_core!(self.0, core, { Ok(core.recv_limit.clone()) })
+    }
+
+    pub async fn send_evt(&self, evt: DataChannelEvent) -> Result<()> {
+        data_chan_weak_core!(self.0, core, { Ok(core.evt_send.clone()) })?
+            .send(evt)
+            .await
     }
 }
 
 /// A go pion webrtc DataChannel.
-#[derive(Debug)]
-pub struct DataChannel(pub(crate) usize);
+pub struct DataChannel(Arc<Mutex<std::result::Result<DataChanCore, Error>>>);
 
-impl Drop for DataChannel {
-    fn drop(&mut self) {
-        unregister_data_chan_evt_cb(self.0);
-        unsafe { API.data_chan_free(self.0) }
+impl std::fmt::Debug for DataChannel {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let res = match &*self.0.lock().unwrap() {
+            Ok(core) => format!("DataChannel(open, {})", core.data_chan_id),
+            Err(err) => format!("DataChannel(closed, {:?})", err),
+        };
+        f.write_str(&res)
     }
 }
 
 impl DataChannel {
+    pub(crate) fn new(
+        data_chan_id: usize,
+        recv_limit: Arc<tokio::sync::Semaphore>,
+    ) -> (Self, EventRecv<DataChannelEvent>) {
+        let (evt_send, evt_recv) = EventSend::new(1024);
+
+        let strong = Arc::new(Mutex::new(Ok(DataChanCore::new(
+            data_chan_id,
+            recv_limit,
+            evt_send.clone(),
+        ))));
+
+        let weak = WeakDataChan(Arc::downgrade(&strong));
+
+        register_data_chan(data_chan_id, weak);
+
+        // we might have missed some state callbacks
+        if let Ok(ready_state) =
+            unsafe { API.data_chan_ready_state(data_chan_id) }
+        {
+            // this is a terrible suggestion clippy
+            #[allow(clippy::comparison_chain)]
+            if ready_state == 2 {
+                let permit = evt_send.try_permit().unwrap();
+                let _ = evt_send.send_permit(DataChannelEvent::Open, permit);
+            } else if ready_state > 2 {
+                let permit = evt_send.try_permit().unwrap();
+                let _ = evt_send.send_permit(DataChannelEvent::Close, permit);
+            }
+        }
+
+        (Self(strong), evt_recv)
+    }
+
+    /// Close this data channel.
+    pub fn close(&self, err: Error) {
+        let mut tmp = Err(err.clone());
+
+        {
+            let mut lock = self.0.lock().unwrap();
```

Comment:
added the swap check here: https://github.com/holochain/tx5/pull/77/commits/6cc8ace3bcaa537e181bb8ebad1ac94a024a5224

I don't disagree about the unit tests, except it's super hard to write tests for this low-level crate, because there are so many fiddly bits, you end up having to duplicate a ton of the connection management code in the higher-level tx5 crate. The main current test is in the lib.rs of this crate.